### PR TITLE
fix: Allow setting Textbox "value" dynamically

### LINKED ISSUE #1578: Text box value clearing
#### Wave SDK Version, OS

0.21.1

#### Actual behavior

I want to clear the value of a textbox when dropdown is changed in a dialog. But this is weirdly interconnected with textbox tooltip. 

Scenario 1: Without tooltip - Not working

```python
from h2o_wave import app, main, Q, ui

@app("/")
async def serve(q: Q) -> None:
    if not q.client.initialized:
        q.page['meta'] = ui.meta_card(box='')
        q.page['example'] = ui.form_card(box='1 1 2 1', items=[
            ui.button(name='show_dialog', label='Open dialog', primary=True)
        ])
        q.client.initialized = True
        q.client.dropdown1 = "1"
    else:
        if q.args.show_dialog:
            show_dialog(q)

        if q.args.dropdown1:
            if q.args.dropdown1 != q.client.dropdown1:
                q.client.dropdown1 = q.args.dropdown1
                q.client.textbox = ""
                show_dialog(q)

        if q.args.cancel:
            q.page['meta'].dialog = None
            q.client.dropdown1 = None
            q.client.textbox = None

    await q.page.save()


def show_dialog(q):
    q.page["meta"].dialog = ui.dialog(
        title="Dialog",
        items=[
            ui.dropdown(
                name="dropdown1",
                label="Dropdown1",
                value=q.client.dropdown1,
                trigger=True,
                choices=[
                    ui.choice(name="1", label="Choice 1"),
                    ui.choice(name="2", label="Choice 2")
                ],
            ),
            ui.textbox(
                name="textbox",
                label="Text box",
                value=q.client.textbox,
            ),
            ui.button(name='cancel', label='Cancel')
        ],
        events=["dismissed"],
        closable=True,
    )

```

https://user-images.githubusercontent.com/29954570/183548744-aaa8c616-1b10-44c3-b227-d8962f71ba55.mov

Scenario 2: With tooltip - Not working

```python
from h2o_wave import app, main, Q, ui

@app("/")
async def serve(q: Q) -> None:
    if not q.client.initialized:
        q.page['meta'] = ui.meta_card(box='')
        q.page['example'] = ui.form_card(box='1 1 2 1', items=[
            ui.button(name='show_dialog', label='Open dialog', primary=True)
        ])
        q.client.initialized = True
        q.client.dropdown1 = "1"
    else:
        if q.args.show_dialog:
            show_dialog(q)

        if q.args.dropdown1:
            if q.args.dropdown1 != q.client.dropdown1:
                q.client.dropdown1 = q.args.dropdown1
                q.client.textbox = ""
                show_dialog(q)

        if q.args.cancel:
            q.page['meta'].dialog = None
            q.client.dropdown1 = None
            q.client.textbox = None

    await q.page.save()


def show_dialog(q):
    q.page["meta"].dialog = ui.dialog(
        title="Dialog",
        items=[
            ui.dropdown(
                name="dropdown1",
                label="Dropdown1",
                value=q.client.dropdown1,
                trigger=True,
                choices=[
                    ui.choice(name="1", label="Choice 1"),
                    ui.choice(name="2", label="Choice 2")
                ],
            ),
            ui.textbox(
                name="textbox",
                label="Text box",
                value=q.client.textbox,
                tooltip="Tooltip"
            ),
            ui.button(name='cancel', label='Cancel')
        ],
        events=["dismissed"],
        closable=True,
    )
```

https://user-images.githubusercontent.com/29954570/183548927-8fe94a08-90ee-402d-aebd-3c7c236b97e0.mov

Scenario 3: With tooltip only for one choice- Working

```python
from h2o_wave import app, main, Q, ui

@app("/")
async def serve(q: Q) -> None:
    if not q.client.initialized:
        q.page['meta'] = ui.meta_card(box='')
        q.page['example'] = ui.form_card(box='1 1 2 1', items=[
            ui.button(name='show_dialog', label='Open dialog', primary=True)
        ])
        q.client.initialized = True
        q.client.dropdown1 = "1"
    else:
        if q.args.show_dialog:
            show_dialog(q)

        if q.args.dropdown1:
            if q.args.dropdown1 != q.client.dropdown1:
                q.client.dropdown1 = q.args.dropdown1
                q.client.textbox = ""
                show_dialog(q)

        if q.args.cancel:
            q.page['meta'].dialog = None
            q.client.dropdown1 = None
            q.client.textbox = None

    await q.page.save()


def show_dialog(q):
    q.page["meta"].dialog = ui.dialog(
        title="Dialog",
        items=[
            ui.dropdown(
                name="dropdown1",
                label="Dropdown1",
                value=q.client.dropdown1,
                trigger=True,
                choices=[
                    ui.choice(name="1", label="Choice 1"),
                    ui.choice(name="2", label="Choice 2")
                ],
            ),
            ui.textbox(
                name="textbox",
                label="Text box",
                value=q.client.textbox,
                tooltip="Choice 1" if q.client.dropdown1 == "1" else None
            ),
            ui.button(name='cancel', label='Cancel')
        ],
        events=["dismissed"],
        closable=True,
    )
```

https://user-images.githubusercontent.com/29954570/183549101-7f9a418b-1ba3-4971-b73e-b866bbfbcb90.mov

Also with this scenario and making tetxbox `trigger=true`,  I found that even though the value of the textbox is empty it will be available as an arg.

https://user-images.githubusercontent.com/29954570/183550904-273c15c0-0054-481f-a1b4-908beba01310.mov

#### Expected behavior

Clear textbox value without any problem.

### PR DESCRIPTION
**Pull request by aalencar**
* allow Textbox "value" to be set from wave app
* add tests for TextField and MaskedTextField
* fix warning about act() in tests

https://user-images.githubusercontent.com/15820796/186417390-982c280e-4dff-44cb-a056-2d86d76bace6.mov

closes #1578

### CODE REVIEW

**Review by mturoci — CHANGES_REQUESTED** [review]
Thanks @aalencar, looking good overall, just a few small comments from me.

Also can you please extract as part of this PR (unrelated refactor).

```ts
const pushMock = jest.fn()
wave.push = pushMock
```

to the top of the test file so that we create pushMock just once and not per test case?

File: ui/src/textbox.test.tsx [hunk 1]
```diff
@@ -12,19 +12,21 @@
...
 const name = 'textbox'
-const textboxProps: Textbox = { name }
+let textboxProps: Textbox = { name }
 
 describe('Textbox.tsx', () => {
   beforeEach(() => {
     jest.clearAllMocks()
     jest.useFakeTimers()
     wave.args[name] = null
+    textboxProps = { name }
```
[the remark this answers is not in the document]
Comment (aalencar) [reply ui/src/textbox.test.tsx:32]: Added.

File: ui/src/textbox.test.tsx [hunk 2]
```diff
@@ -51,7 +53,7 @@ describe('Textbox.tsx', () => {
   it('Sets args on input', () => {
     const { getByTestId } = render(<XTextbox model={textboxProps} />)
     fireEvent.change(getByTestId(name), { target: { value: 'text' } })
-    jest.runOnlyPendingTimers()
+    act(() => { jest.runOnlyPendingTimers() })
```
[the remark this answers is not in the document]
Comment (aalencar) [reply ui/src/textbox.test.tsx]: We should use act when we change the state outside of React call stack.
The warning is not showing up for me anymore. Reverted.

Related: https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning

File: ui/src/textbox.test.tsx [hunk 3]
```diff
@@ -79,10 +81,26 @@ describe('Textbox.tsx', () => {
     fireEvent.change(getByTestId(name), { target: { value: 'aaa' } })
 
     expect(pushMock).not.toBeCalled() // Not called immediately, but after specified timeout.
-    jest.runOnlyPendingTimers()
+    act(() => { jest.advanceTimersByTime(250) })
+    expect(pushMock).not.toBeCalled()
+    act(() => { jest.advanceTimersByTime(250) })
```
[the remark this answers is not in the document]
Comment (aalencar) [reply ui/src/textbox.test.tsx]: Agreed. Reverted.

File: ui/src/textbox.tsx [hunk 4]
```diff
@@ -71,20 +71,25 @@ const DEBOUNCE_TIMEOUT = 500
 export const
   XTextbox = ({ model: m }: { model: Textbox }) => {
     const
+      [value, setValue] = React.useState(m.value ?? ''),
+      debounceRef = React.useRef(debounce(DEBOUNCE_TIMEOUT, () => wave.push())),
```
[the remark this answers is not in the document]
Comment (mturoci) [reply ui/src/textbox.tsx]: You are correct, thanks for pointing out.

**Pull request merged**